### PR TITLE
Add gradient descent docs - bump Julia compat to 1.10

### LINKED ISSUE #51: Examples of fitting with gradient descent
- with loglikelihood
- inside EM

### PR DESCRIPTION
**Pull request by rsenne**
Adds a documentation example to resolve #51. Adds two examples 1.) fitting an hmm inside EM with GD and 2.) fitting an HMM by using autodiff + GD over all parameters. Let me know how this looks! happy to make changes

---

EDIT by @gdalle:

- Bumped Julia lower bound to 1.10
- Bumped package version to 0.7.1

### CODE REVIEW

File: examples/gradientdescent.jl [hunk 1]
```diff
@@ -0,0 +1,236 @@
+# # Gradient Descent in HMMs
+
+#= 
+In this tutorial we explore two ways to use gradient descent when fitting HMMs:
+
+1. Fitting parameters of an observation model that do not have closed-form updates
+   (e.g., GLMs, neural networks, etc.).
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
1. Fitting parameters of an observation model that do not have closed-form updates
   (e.g., GLMs, neural networks, etc.), inside the EM algorithm.
```

File: examples/gradientdescent.jl [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+We begin by defining a Normal observation model.
+=#
+
+mutable struct NormalModel{T}
+    μ::T
+    logσ::T  # unconstrained parameterization; σ = exp(logσ)
+end
+
+stddev(mod::NormalModel) = exp(mod.logσ)
```
Comment (gdalle) [inline examples/gradientdescent.jl:43]:
```suggestion
mean(mod::NormalModel) = mod.μ
stddev(mod::NormalModel) = exp(mod.logσ)
```

File: examples/gradientdescent.jl [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+log of the standard deviation. Using logσ is intentional so we can optimize over
+all real numbers without worrying about the positivity constraint on σ.
+
+Next, we provide the minimal interface expected by HiddenMarkovModels.jl:
+(logdensityof, rand, fit!).
+=#
+
+function DensityInterface.logdensityof(mod::NormalModel, obs::T) where {T<:Real}
+    s = stddev(mod)
+    return -0.5 * log(2π) - log(s) - 0.5 * ((obs - mod.μ) / s)^2
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
    return - log(2π) / 2 - log(s) - ((obs - mean(mod)) / s)^2 / 2
```
Dividing by the integer `2` preserves single precision if needed

File: examples/gradientdescent.jl [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+end
+
+DensityInterface.DensityKind(::NormalModel) = DensityInterface.HasDensity()
+
+function Random.rand(rng::AbstractRNG, mod::NormalModel)
+    return rand(rng, Normal(mod.μ, stddev(mod)))
+end
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
function Random.rand(rng::AbstractRNG, mod::NormalModel{T}) where {T}
    return stddev(mod) * randn(rng, T) + mean(mod)
end
```
That way we don't need Distributions.jl at all

File: examples/gradientdescent.jl [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+1. Fitting parameters of an observation model that do not have closed-form updates
+   (e.g., GLMs, neural networks, etc.).
+2. Fitting the entire HMM with gradient-based optimization by leveraging automatic
+   differentiation.
+
+We will explore both approaches below.
+=#
+
+using DensityInterface
+using Distributions
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
```
Since we implement our own distribution, we shouldn't need this

File: examples/gradientdescent.jl [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+stddev(mod::NormalModel) = exp(mod.logσ)
+
+#= 
+We have defined a simple probability model with two parameters: the mean and the
+log of the standard deviation. Using logσ is intentional so we can optimize over
+all real numbers without worrying about the positivity constraint on σ.
+
+Next, we provide the minimal interface expected by HiddenMarkovModels.jl:
+(logdensityof, rand, fit!).
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
`(logdensityof, rand, fit!)`.
```

File: examples/gradientdescent.jl [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    μ::T
+    logσ::T  # unconstrained parameterization; σ = exp(logσ)
+end
+
+stddev(mod::NormalModel) = exp(mod.logσ)
+
+#= 
+We have defined a simple probability model with two parameters: the mean and the
+log of the standard deviation. Using logσ is intentional so we can optimize over
+all real numbers without worrying about the positivity constraint on σ.
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
log of the standard deviation. Using `logσ` is intentional so we can optimize over
all real numbers without worrying about the positivity constraint on `σ`.
```

File: examples/gradientdescent.jl [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+function Random.rand(rng::AbstractRNG, mod::NormalModel)
+    return rand(rng, Normal(mod.μ, stddev(mod)))
+end
+
+#= 
+Because we are fitting a Gaussian (and the variance can collapse to ~0), we add
+weak priors to regularize the parameters. We use:
+- A weak Normal prior on μ
+- A moderate-strength Normal prior on logσ that pulls σ toward ~1
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
- A weak Normal prior on `μ`
- A moderate-strength Normal prior on `logσ` that pulls `σ` toward ~1
```

File: examples/gradientdescent.jl [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+=#
+
+const μ_prior = Normal(0.0, 10.0)
+const logσ_prior = Normal(log(1.0), 0.5)
```
Comment (gdalle) [inline examples/gradientdescent.jl]: Can we use `NormalModel` for these too, since we already have the logpdf?

File: examples/gradientdescent.jl [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+)
+    function neglogpost(θ)
+        μ, logσ = θ
+
+        #- Weighted negative log-likelihood
+        nll = 0.0
+        tmp = NormalModel(μ, logσ)
+        for (y, w) in zip(data, weights)
+            nll += -w * logdensityof(tmp, y)
+        end
```
Comment (gdalle) [inline examples/gradientdescent.jl]: Perhaps using a `mapreduce` would be more efficient for Zygote users

File: examples/gradientdescent.jl [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+const logσ_prior = Normal(log(1.0), 0.5)
+
+function StatsAPI.fit!(
+    mod::NormalModel, data::AbstractVector{<:Real}, weights::AbstractVector{<:Real}
+)
+    function neglogpost(θ)
+        μ, logσ = θ
+
+        #- Weighted negative log-likelihood
+        nll = 0.0
```
Comment (gdalle) [inline examples/gradientdescent.jl]: Use a `zero` of the correct type?

File: examples/gradientdescent.jl [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+- A moderate-strength Normal prior on logσ that pulls σ toward ~1
+=#
+
+const μ_prior = Normal(0.0, 10.0)
+const logσ_prior = Normal(log(1.0), 0.5)
+
+function StatsAPI.fit!(
+    mod::NormalModel, data::AbstractVector{<:Real}, weights::AbstractVector{<:Real}
+)
+    function neglogpost(θ)
```
Comment (gdalle) [inline examples/gradientdescent.jl]: I would define this `neglogpost(mod, data, weights)` function outside of `fit!` and then only define a closure over constant arguments inside `fit!`

File: examples/gradientdescent.jl [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+obs_dists = [
+    NormalModel(-3.0, log(0.25)), NormalModel(0.0, log(0.5)), NormalModel(3.0, log(0.75))
+]
+
+hmm_true = HMM(init_dist, init_trans, obs_dists)
+
+#= 
+We can now generate data from this HMM.
+Note: rand(rng, hmm, T) returns (state_seq, obs_seq).
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
Note: `rand(rng, hmm, T)` returns `(state_seq, obs_seq)`.
```

File: examples/gradientdescent.jl [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+function of all model parameters.
+
+We can therefore optimize the negative log-likelihood returned by `forward`.
+Each objective evaluation runs the forward algorithm, which can be expensive for
+large datasets, but this approach allows end-to-end gradient-based fitting for
+arbitrary parameterized HMMs.
+
+To respect HMM constraints, we optimize unconstrained parameters and map them to
+valid probability distributions via softmax:
+- π = softmax(ηπ)
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
- `π = softmax(ηπ)`
```

File: examples/gradientdescent.jl [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+- each row of A = softmax(row logits)
```
Comment (gdalle) [inline examples/gradientdescent.jl]:
```suggestion
- each row of `A` = `softmax(row_logits)`
```

File: examples/gradientdescent.jl [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+
+function rowsoftmax(M::AbstractMatrix)
+    A = similar(M)
+    for i in 1:size(M, 1)
+        A[i, :] .= softmax(view(M, i, :))
+    end
+    return A
+end
+
+function unpack_to_hmm(θ::AbstractVector, K::Int)
```
Comment (gdalle) [inline examples/gradientdescent.jl]: Using a `ComponentVector` here would be much less painful

Comment (gdalle) [reply examples/gradientdescent.jl]: Also, you wouldn't need `K`

File: examples/gradientdescent.jl [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+
+# Convert a valid HMM into an unconstrained parameter vector θ
+function hmm_to_θ0(hmm::HMM)
+    K = length(hmm.init)
+
+    π = hmm.init
+    A = hmm.trans
+
+    ηπ = log.(π .+ eps()) # Logits + eps() for numerical stability
+    ηA = log.(A .+ eps())
```
Comment (gdalle) [inline examples/gradientdescent.jl]: Use the correct number type inside `eps`

**Review by gdalle — CHANGES_REQUESTED** [review]
I like the idea very much! Here are a few suggestions, thank you for the initiative

**Comment by rsenne** [comment]
Test for current version of Julia failed as `Optim` has API changes in recent breaking version `2.0.0`. I changed tutorial to reflect this new API. All other tests pass locally

**Comment by gdalle** [comment]
I told JET to chill but I think there is still an error somewhere in the docs build?

**Comment by rsenne** [comment]
Yes, I tried to fix it but for some reason it works locally but not here. I'm admittedly a bit stumped as the error message here is a bit unhelpful. I'll try to do more sleuthing today

**Comment by gdalle** [comment]
Might be Zygote acting up on Julia 1.12? In which case the switch to 1.11 will help

**Comment by rsenne** [comment]
i think i found the issue. i accidentally introduced spaces between `#` and `src`which was leading to an issue with Literate

**Review by gdalle — APPROVED** [review]
